Sources/DequeModule/Compatibility.swift

```diff
@@ -0,0 +1,60 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Collections open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension Array {
+  /// Returns true if `Array.withContiguousStorageIfAvailable` is broken
+  /// in the stdlib we're currently running on.
+  ///
+  /// See https://bugs.swift.org/browse/SR-14663.
+  @inlinable
+  internal static func _isWCSIABroken() -> Bool {
+    guard _isBridgedVerbatimToObjectiveC(Element.self) else {
+      // SR-14663 only triggers on array values that are verbatim bridged
+      // from Objective-C, so it cannot ever trigger for element types
+      // that aren't verbatim bridged.
+      return false
+    }
+
+    // SR-14663 was introduced in Swift 5.1. Check if we have a broken stdlib.
+
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    // The bug is caused by a bogus precondition inside a non-inlinable stdlib
+    // method, so to determine if we're affected, we need to check the currently
+    // running OS version.
+    guard #available(macOS 10.15, iOS 13, watchOS 6, tvOS 13, *) else {
+      // The OS is too old to be affected by this bug.
+      return false
+    }
+    // FIXME: When a stdlib is released that contains a fix, add a check for it.
+    return true
+#else
+    // On non-ABI-stable platforms, we are always affected, because this
+    // package only supports Swift 5.3 or later.
+    // FIXME: When a stdlib is released that contains a fix, add a check for it.
+    return true
+#endif
+  }
+}
+
+extension Sequence {
+  // An adjusted version of the standard `withContiguousStorageIfAvailable`
+  // method that works around https://bugs.swift.org/browse/SR-14663.
+  @inlinable
+  internal func _withContiguousStorageIfAvailable_SR14663<R>(
+    _ body: (UnsafeBufferPointer<Element>) throws -> R
+  ) rethrows -> R? {
+    if Self.self == Array<Element>.self && Array<Element>._isWCSIABroken() {
+      return nil
```

Comment:
If we had a reliable/easy way to test if a particular array instance is bridged, then we could let native arrays pass through here. 🤔

Reply:
Maybe:

```swift
if let array = self as? Array<Element>, array._baseAddressIfContiguous == nil { return nil }
```

Tested in a Playground:
```swift
import Foundation
var a: [NSObject] = [NSObject()]
print(a._baseAddressIfContiguous) // Optional(0x00007fdf7e40c0d0)

a = NSArray(array: [NSObject()]) as! [NSObject]
print(a._baseAddressIfContiguous) // nil
```

Reply:
Huh, that's a good find. Looking at similarly exposed members, `Array._hoistableIsNativeTypeChecked()` might be even better.

There are (at least) two worries with using these underscored APIs: (including `_isBridgedVerbatimToObjectiveC` above)

1. Swift doesn't promise it won't change their behavior, so it's tricky to safely use them in code that is not part of the stdlib. E.g., I imagine for element types that don't require a dynamic type check at every access, we may want to change `_baseAddressIfContiguous` to return non-nil for bridged arrays that use specific known NSArray subclasses. 

2. These have also been made public for historical reasons, rather than any intentional design, and we may find a good reason to change them to `@inlinable internal`, disallowing direct calls like this in future SDKs. We wouldn't be able to remove the entry points themselves (they are part of Swift's ABI), but we might need to restrict them to only be called from inlinable functions within the stdlib.

As this package is maintained by the same people who work on the Swift stdlib, I think it's not unreasonable for this package to use these. At the first sign of trouble, we can remove the calls, or limit them to older Swift toolchains. The primary danger is that this code will spread beyond this package through vendoring or uncritical copy-n-paste. 

My instinct as a stdlib engineer tells me that the risk is low with `_hoistableIsNativeTypeChecked` (and pretty much nonexistent for `_isBridgedVerbatimToObjectiveC`); on the other hand, this package is inherently averse to even low-risk stuff. Hm.

Reply:
I think the best thing to do is to keep this as is for now until we have a Swift release (let's call it x.y) that contains a fix for the underlying issue. In the (unlikely) case that the fix won't back deploy, then we can add a call to `_hoistableIsNativeTypeChecked` restricted to only be called on Swift 5.1 ..< x.y.

Reply:
I think that's fine. I would also be OK with doing the workaround here now, so long as we put a comment in the stdlib that references this bug.